### DISCUSSION TITLE
Commits by different users on a shared machine and repository

### DISCUSSION THREAD

**Discussion by gsaponaro** [robotology/community#77] [Q&A] [answered]
Hi, we would like to be able to make git commits from a machine that is shared among many users in our lab (iCub laptop). Each user has a personal GitHub account. For example, today I'd like to make a commit "signed by me", tomorrow a colleague would make another by him, etc., and all of this would happen on the single clone of the repository that exists on the machine.

We thought about changing the global file `~/.gitconfig` before every commit, but it's not the most practical solution. Is there a better way?

In the icub-main history, I noticed that sometimes the phrase "committed with" appears, like shown below. Perhaps that's what we'd like to achieve.

![github_multiple_users](https://cloud.githubusercontent.com/assets/6716962/11239351/419900b8-8de3-11e5-9bdf-4cb4269740d5.png)

**Comment by Tobias-Fischer** [comment]
Hi,

for example https://stackoverflow.com/questions/18750808/difference-between-author-and-committer-in-git and https://git-scm.com/docs/git-commit (scroll to --author).

Best, Tobias

**Comment by drdanz** [comment, marked as answer]
If you are using different repositories you can also set the name and email in just for a specific repository, but it's not a very clean solution, and I don't know if it is the case here.

You can use `git commit --author="John Doe <john.doe@example.com>"` to change the author when you make a commit.

Another option is to use the environment variables:

```
GIT_AUTHOR_NAME
GIT_AUTHOR_EMAIL
GIT_COMMITTER_NAME
GIT_COMMITTER_EMAIL
```

"Committer" is the "iCub User" in the picture you added.

**Comment by gsaponaro** [comment]
Thanks a lot. I applied the suggestions and also created Bash aliases to avoid typing the `--author` option every time:

```
GIT_COMMITTER_NAME="iCubLisboa team"
GIT_COMMITTER_EMAIL="username@users.noreply.github.com"
alias git-commit-as-gsaponaro="git commit --author=\"Giovanni Saponaro <gsaponaro@isr.ist.utl.pt>\""
alias git-commit-as-anotherUser=...
```

**Comment by pattacini** [comment]
@drdanz do you know what will happen for a commit where `email` and `author` options do not match each other?

**Comment by pattacini** [comment]
Uhm ... it seems that somehow the information conveyed by the `email` field is quite relevant for GitHub:

https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user

**Comment by pattacini** [comment]
I've noticed now that the alias does contain the `email` too; hence there won't be any mismatch.

**Comment by drdanz** [comment]
@pattacini i can confirm that github uses the author _email_ to associate a commit to one user, or the author _name_ if the email does not correspond to any user.
At a "git" level you will see both the _name_ and the _email_ in the log.

**Comment by drdanz** [comment]
Closing this, @gsaponaro please reopen the issue if you need more information.

**Comment by gsaponaro** [comment]
Perfect, thanks.
